updates for new 122 structure (#14)
cse12x/java-tutorial: book_source/source/2darrays.md

# <i class="fas fa-book fa-fw"></i> 2DArrays

## 2DArray Fundamentals

> An 2D array is very similar to an 1D array, or just an array as we have learned, only that instead of storing primiative types, it stores another array. In other words, a 2D array is an array where each element contains another array. Commonly this is represented/used when dealing with images, graphs, tables, and various other 2D representations.

Reviewer: Reworking to clean up some grammar and simplify. 
```suggestion
A 2D or "two-dimensional" array is very similar to the arrays that we talked about in the previous section, but instead of elements of types like `int`, `double`, `String`, etc, its element type is _array_. In other words, a 2D array is an array where each element is a reference to _another_ array. This type of data structure is helpful for representing many different kinds of scenarios, such as in applications using images, graphs, and tables. 
```

> 

Reviewer: I think introducing the idea of "rows" vs. "columns" would make more sense to come here, before we get into syntax!

> Much like how you would access a specific element in an array using square brackets '[]', for a 2D array, you need to add one more pair of brackets to access an element in a 2D array. For example, if you wanted to create a 2D array storing integers, an example syntax could be int[][] (read as a 2D array of ints or 2D int array). Although the specific lengths of the arrays can vary, these arrays still have a fixed length once created. For ease of representation, we typically repersent a 2D array with rows being the inner arrays and columns being the array of arrays. For example:

Reviewer: ```suggestion
You can access elements of a 2D array using square brackets `[]`, in the same way that we accessed elements of 1D arrays in the previous section. However, remember that the elements of a 2D array are arrays! If you want to get one of _those_ arrays' elements, you'll need to use two sets of square brackets `[][]`. For example, if you were dealing with a 2D array of integers `arr`, the type of the 2D array would be `int[][]` and you could access an individual `int` from the 2D array with `arr[_i_][_j_]` where `i` and `j` are the row and column index of the element we wanted. 
```

> Much like how you would access a specific element in an array using square brackets '[]', for a 2D array, you need to add one more pair of brackets to access an element in a 2D array. For example, if you wanted to create a 2D array storing integers, an example syntax could be int[][] (read as a 2D array of ints or 2D int array). Although the specific lengths of the arrays can vary, these arrays still have a fixed length once created. For ease of representation, we typically repersent a 2D array with rows being the inner arrays and columns being the array of arrays. For example:

Reviewer: While we could mention jagged 2D arrays, I don't think it's worth getting into -- they're not really covered in depth in 12x and we'll never ask them to deal with them.

Reviewer 2: Probably fine in this version now that it's only mentioned in passing.

```
[[0 , 1 , 2 , 3 , 4],
[5 , 6 , 7 , 8 , 9 ],
[10, 11, 12, 13, 14],
[15, 16, 17, 18, 19],
[20, 21, 22, 23, 24]]
```

> Notice there is an extra pair of opening and closing brackets, which represents the array that stores all of the other arrays. The syntax is commonly represented as array[row][col] and is represented visually as so above for clarity. Additionally, the call to an array using the syntax array[1] would still function, only instead of providing a primative type, it would instead provide the array stored at that index of the array of arrays.

Reviewer: I think we can drop this paragraph -- the same general information has been covered already.

Reviewer 2: I think it's fine to reinforce this here.

````{tab-set-code}

```{code-block} java
int[][] array = new int[5][5];
array[1][2] = 1;
array[1][3] = 14;
array[3][4] = array[1][2];
array[4][4] = 5;

for (int i = 0; i < array.length; i++) {
    System.out.println(Arrays.toString(array[i]));
}
```

```{code-block} python
array = [[0]*5] *5  # Short-hand to make a 2D length 5 list
array[1][2] = 1;
array[1][3] = 14;
array[3][4] = array[1][2];
array[4][4] = 5;

print(array)
```

```{code-block} javascript
let array = [[0,0,0,0,0], [0,0,0,0,0], [0,0,0,0,0], [0,0,0,0,0], [0,0,0,0,0]];

array[1][2] = 1;
array[1][3] = 14;
array[3][4] = array[1][2];
array[4][4] = 5;

console.log(array);
```

```{code-block} c

> int array[5][5] = {{0}};

Reviewer: ```suggestion
int array[5][5];
```
I think this is a bit more "typical" C-style

array[1][2] = 1;
array[1][3] = 14;
array[3][4] = array[1][2];
array[4][4] = 5;

> printf("[");

Reviewer 2: Needs indentation in this loop and I don't think the output formatting is correct (trailing commas, no brackets around inner arrays and no closing bracket)

> printf("[");
> for (int i = 0; i < 5; i++) {
> for (int j = 0; j < 5; j++) {
> printf("%d, ", array[i][j]);
> }
> printf("\n");
> }

Reviewer: ```suggestion
printf("[");
for (int i = 0; i < 5; i++) {
    for (int j = 0; j < 5; j++) {
        printf("%d, ", array[i][j]);
    }
    printf("\n");
}
```
fix indentation

```

```{code-block} c++

> int array[5][5] = {
>     {0, 0, 0, 0, 0},
>     {0, 0, 0, 0, 0},
>     {0, 0, 0, 0, 0},
>     {0, 0, 0, 0, 0},
>     {0, 0, 0, 0, 0},
> };

Reviewer: ```suggestion
int array[5][5];
```
ditto but C++-style

array[1][2] = 1;
array[1][3] = 14;
array[3][4] = array[1][2];
array[4][4] = 5;
for (int i = 0; i < 5; ++i) {
for (int j = 0; j < 5; ++j) {

> std::cout << array[i][j] << " ";

Reviewer 2: Needs indentation and I think you should put the `endl` on each line of the inner loop too.

> for (int i = 0; i < 5; ++i) {
> for (int j = 0; j < 5; ++j) {
> std::cout << array[i][j] << " ";
> }
> std::cout << std::endl; 
> }

Reviewer: ```suggestion
for (int i = 0; i < 5; ++i) {
    for (int j = 0; j < 5; ++j) {
        std::cout << array[i][j] << " ";
    }
    std::cout << std::endl; 
}
```

```
````

> Similarily, if we want to change the value at (3,4) of the 2D array to 1, where 3 is the "x" coordinate or outside array and 4 is the "y" or inner array we would do something like array[3][4] = 1. Note that the 2D array can only be of one type, just like how a 1D array is only one type.

Reviewer: ```suggestion
Note the syntax that we are using to assign values to integer elements of the 2D array - we need to use two pairs of square brackets `[][]` and specify the row _and_ column of the space we want to assign the value to. 
```
Simplified because some of the same general information is covered earlier.

## Looping over 2D Arrays

> We can also use a for loop to view and/or modify each index, although since this is a nested data structure (meaning there is a data structure within another data structure), we need to use a nested for loop in return. The starting value for each loop will typically start at 0, going to either array.length or array[0].length. Note the second for loop, we use array[0].length since we only want to loop through to the end of the inner array length, represented by array[0].

Reviewer: Shortening this a bit - I think the walkthrough of the initialization value and tests isn't strictly necessary, since it's evident in the code right below. 
```suggestion
To traverse over a 2D array, it's very common to use a _nested for loop structure_ where the outer for loops iterates over each _row_ of the 2D array, and the inner for loop iterates over each _element_ within each row. See some examples below. 
```

````{tab-set-code}

```{code-block} java
int[][] array = new int[5][5];
// Code to fill array with some values

int sum = 0;
for (int i = 0; i < array.length; i++) {
    for (int j = 0; j < array[0].length; j++) {
        sum += array[i][j];
    }
}
```

> ```{code-block} python

Reviewer 2: All of the code examples in this section are still 1D arrays

array = [0] * 5  # Short-hand to make a length 5 list
# Code to fill array with some values

sum = 0
for i in range(len(array)):
    sum += array[i]
```

```{code-block} javascript
let array = [0, 0, 0, 0, 0];
// Code to fill array with some values

let sum = 0;
for (let i = 0; i < array.length; i++) {
    sum += array[i];
}
```

```{code-block} c
int array[5];
// Code to fill array with some values

int sum = 0;
size_t length = sizeof(array) / sizeof(array[0]);
for (int i = 0; i < length; i++) {
    sum += array[i];
}
```

```{code-block} c++
int array[5];
// Code to fill array with some values

int sum = 0;
size_t length = sizeof(array) / sizeof(array[0]);
for (int i = 0; i < length; i++) {
    sum += array[i];
}
```

```{code-block} c++
int array[5];
// Code to fill array with some values

int sum = 0;
for (int i = 0; i < array.length; i++) {
    sum += array[i];
}
```
````

> ## Printing 2D Arrays

Reviewer 2: I think it seems a bit redundant to have this section when it also has all of the printing code in the intro section. I think it maybe clearer if there was a way to remove all the looping code from earlier to print if possible.

> Note that unlike with a 1D array, we cannot simply do Arrays.toString(array) on the entire 2D array, instead we have to loop thorugh each of the rows of the 2D array, then call Arrays.toString(array[i]), where i are values 0 through array.length - 1.

Reviewer: A bit of clean up + addition of `Arrays.deepToString(...)`
```suggestion
Recall that we would get a nicely-formatted String representation of a 1D array using `Arrays.toString(...)`. However, passing a 2D array to this method will probably not produce the output you're looking for. Instead, you should use the `Arrays.**deep**ToString(...)` method or iterate over each of the rows of the 2D array, then print out each individual row using `Arrays.toString(...)` as shown below. 
```

````{tab-set-code}

```{code-block} java

int[][] array = new int[5][5];
array[1][2] = 1;
array[1][3] = 14;
array[3][4] = array[1][2];
array[4][4] = 5;

for (int i = 0; i < array.length; i++) {
    System.out.println(Arrays.toString(array[i]));
}
```

```{code-block} python
array = [0] * 3  # Short-hand to make a length 3 list
array[0] = 8
array[1] = 1
array[2] = 7

print(array)
// Output: [8, 1, 7]
```

```{code-block} javascript
let array = [[], [], [], []];

> 

Reviewer: Are we missing the step of actually putting values into the array?

Reviewer 2: This is still unresolved @WateringFire

console.log(array);
// Output: [8, 1, 7]
```

```{code-block} c
int array[3];
array[0] = 8;
array[1] = 1;
array[2] = 7;

> printf("[");
> for (int i = 0; i < 2; i++) {
>     printf("%d, ", array[i]);
> }
> printf("%d]\n", array[2]);

Reviewer: ```suggestion
printf("[");
for (int i = 0; i < 3; i++) {
    printf("%d, ", array[i]);
}
printf("]\n");
```

// Output: [8, 1, 7]
```

```{code-block} c++
int array[3];
array[0] = 8;
array[1] = 1;
array[2] = 7;

> cout << "[";
> for (int i = 0; i < 2; i++) {
>     cout << array[i] << ", ";
> }
> cout << array[2] << "]" << endl;

Reviewer: Add `std` namespace qualification (since it's used in C++ examples above), remove unnecessary fenceposting
```suggestion
std::cout << "[";
for (int i = 0; i < 3; i++) {
    cout << array[i] << ", ";
}
std::cout << "]" << endl;
```
